be/src/exprs/bitmap_function.cpp

```diff
@@ -22,6 +22,256 @@
+}
+
+template<>
+int32_t serialize_size(const DecimalV2Value& v) {
+    return DECIMAL_BYTE_SIZE;
+}
+
+template<>
+int32_t serialize_size(const StringValue& v) {
+    return v.len + 4;
+}
+// serialize_size end
+
+// write_to start
+template<typename T>
+char* write_to(const T& v, char* dest) {
+    size_t type_size = sizeof(T);
+    memcpy(dest, &v, type_size);
+    dest += type_size;
+    return dest;
+}
+
+template<>
+char* write_to(const DateTimeValue& v, char* dest) {
+    DateTimeVal value;
+    v.to_datetime_val(&value);
+    *(int64_t*)dest = value.packed_time;
+    dest += DATETIME_PACKED_TIME_BYTE_SIZE;
+    *(int*)dest = value.type;
+    dest += DATETIME_TYPE_BYTE_SIZE;
+    return dest;
+}
+
+template<>
+char* write_to(const DecimalV2Value& v, char* dest) {
+    __int128 value = v.value();
+    memcpy(dest, &value, DECIMAL_BYTE_SIZE);
+    dest += DECIMAL_BYTE_SIZE;
+    return dest;
+}
+
+template<>
+char* write_to(const StringValue& v, char* dest) {
+    *(int32_t*)dest = v.len;
+    dest += 4;
+    memcpy(dest, v.ptr, v.len);
+    dest += v.len;
+    return dest;
+}
+// write_to end
+
+// read_from start
+template<typename T>
+void read_from(const char** src, T* result) {
+    size_t type_size = sizeof(T);
+    memcpy(result, *src, type_size);
+    *src += type_size;
+}
+
+template<>
+void read_from(const char** src, DateTimeValue* result) {
+    DateTimeVal value;
+    value.is_null = false;
+    value.packed_time = *(int64_t*)(*src);
+    *src += DATETIME_PACKED_TIME_BYTE_SIZE;
+    value.type = *(int*)(*src);
+    *src += DATETIME_TYPE_BYTE_SIZE;
+    *result = DateTimeValue::from_datetime_val(value);;
+}
+
+template<>
+void read_from(const char** src, DecimalV2Value* result) {
+    __int128 v = 0;
+    memcpy(&v, *src, DECIMAL_BYTE_SIZE);
+    *src += DECIMAL_BYTE_SIZE;
+    *result = DecimalV2Value(v);
+}
+
+template<>
+void read_from(const char** src, StringValue* result) {
+    int32_t length = *(int32_t*)(*src);
+    *src += 4;
+    *result = StringValue((char *)*src, length);
+    *src += length;
+}
+// read_from end
+
+} // namespace detail
+
+template<typename T>
+struct BitmapRetention {
+public:
+    BitmapRetention() {}
+
+    explicit BitmapRetention(const char* src) {
+        deserialize(src);
+    }
+
+    void add_key(T key) {
+        _keys.push_back(key);
+    }
+
+    void update(const T& key, const RoaringBitmap& bitmap) {
+        typename std::vector<T>::iterator it = std::find(_keys.begin(), _keys.end(), key);
+        if (it != _keys.end()) {
+            if (_bitmaps.find(key) != _bitmaps.end()) {
+                _bitmaps[key].merge(bitmap);
+            } else {
+                _bitmaps[key] = bitmap;
+            }
+        }
+    }
+
+    void merge(const BitmapRetention& other) {
+        // Note, in merge phase, this _keys is empty and other's _keys is whole
+        if (_keys.empty()) {
+            _keys = other._keys;
+        }
+        for(auto& key: _keys) {
```

Comment:
```suggestion
        for (auto& key: _keys) {
```
some other place has the same typo

Reply:
OK